### ISSUE TITLE
unconditional_recursion false-positive in functions without normal return

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#30097]
```rust
enum T {
    A(Box<T>),
    B
}

fn get_ast_type_type(t: &T) {
    match *t {
        T::A(ref tt) => {
            get_ast_type_type(tt);
        },
        T::B => panic!()
    }
}

fn main(){}
```

results in

```
test.rs:6:1: 13:2 warning: function cannot return without recurring, #[warn(unconditional_recursion)] on by default
test.rs: 6 fn get_ast_type_type(t: &T) {
test.rs: 7     match *t {
test.rs: 8         T::A(ref tt) => {
test.rs: 9             get_ast_type_type(tt);
test.rs:10         },
test.rs:11         T::B => panic!()
           ...
test.rs:9:13: 9:34 note: recursive call site
test.rs:9             get_ast_type_type(tt);
                      ^~~~~~~~~~~~~~~~~~~~~
test.rs:6:1: 13:2 help: a `loop` may express intention better if this is on purpose
```

**Comment by arielb1** [comment]
`match` is not involved here. `unconditional_recursion` complains when there is a path in the function that recurs and no path in the function that terminates without recurring. For example:
```Rust
extern "C" { fn rand() -> i32; }

fn q() {
    if unsafe { rand()%2 == 0 } {
        q(); // this path recurs
    } else {
        panic!(); // this path does not terminate
    }
}
fn main() { q(); }
```

**Comment by tbu-** [comment]
Is this a false-positive? You could as well just call `panic!`

**Comment by huonw** [comment]
The lint makes the (slightly) opinionated decision that `panic!()`ing (and other forms of divergence) are not a useful/"correct" way for a function to exit. I believe this is actually required to catch code like the following, which certainly *feels* like it should be considered unconditional recursion:

```rust
fn foo(x: i32) {
    assert!(x < 10);
    foo(x - 1);
}
```

The only thing `get_ast_type_type` or `foo` will do is call themselves some number of times (possibly zero), and then panic; which doesn't seem like the best arrangement.

(FWIW, I worded the error message somewhat carefully to cover this case exactly: it says "cannot return without recurring" rather than "always recurs", since panicking/diverging aren't returns.)

I'm closing as not a bug, but feel free to reopen, if anyone strongly disagrees.

**Comment by nagisa** [comment]
> You could as well just call panic!

You’re focusing on `panic!` too much :)

---

I still think it is wrong to choose to emit or not emit this lint depending on diverge-ness of all other branches. E.g. consider this:

```
enum T {
    A(Box<T>),
    B
}

fn get_ast_type_type(t: &T) {
    match *t {
        T::A(ref tt) => get_ast_type_type(tt),
        T::B => ::std::process::exit(0),
    }
}

fn main(){}
```

which could be totally plausible code.

@huonw gentle reminder that plebeians such as myself can’t reopen issues closed by owners/collaborators.

**Comment by huonw** [comment]
> which could be totally plausible code .

I think it is somewhat implausible; and not really plausible enough to hamstring this lint for other far more common situations. As the lint suggests, things can be rewritten with a loop, which expresses the intention (of looping and exiting without ever doing any real work) better.

> gentle reminder that plebeians such as myself can’t reopen issues closed by owners/collaborators.

The standard protocol is that people without the privileges just ask to reopen (either here or on IRC), sorry for being less clear.

**Comment by nagisa** [comment]
> I think it is somewhat implausible

Okay, I won’t push it hard then, I guess.
